crates/hir-ty/src/tests/traits.rs

```diff
@@ -82,6 +82,46 @@ async fn test() {
     );
 }
 
+#[test]
+fn infer_async_closure() {
+    check_types(
+        r#"
+//- minicore: future, option
+async fn test() {
+    let f = async move |x: i32| x + 42;
+    f;
+//  ^ |i32| -> impl Future<Output = i32>
```

Comment:
This probably shouldn't need `x: i32`, but otherwise it deduces the type as `|i32| -> {unknown}`. It's not an issue with this PR though, and it also happens with non-async closures.